### PR TITLE
Improve `analyzegc` experience

### PR DESCRIPTION
**Pull request by staticfloat**
This makes the `analyzegc` experience faster, more beautiful, and a lot easier to run from a clean clone, as we expect to do regularly on CI.  In particular, the `clang` analysis pass is now invoked in serial, which reduces runtime from 6.5 minutes to 2.5 minutes on a 4-core box; which isn't a bad speedup.

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -323,13 +323,27 @@ clean-support:
 
 cleanall: clean clean-flisp clean-support
 
-analyzegc:
+$(build_shlibdir)/libGCCheckerPlugin.$(SHLIB_EXT): $(SRCDIR)/clangsa/GCChecker.cpp
+	@$(call PRINT_CC, $(CXX) -g -shared -fno-rtti -fPIC -o $@ -std=c++11 -DCLANG_PLUGIN -I$(build_includedir)  -I$(LLVM_SRC_DIR)/include -L$(build_libdir) $< -lclangAnalysis -lclangStaticAnalyzerCore -lclangASTMatchers -lclangAST -lclangLex -lclangBasic)
```
Comment (vtjnash) [inline src/Makefile]: seems like most of these flags are supposed to be coming from `llvm-config`, but got entered manually here

File: src/Makefile [hunk 2]
```diff
@@ -323,13 +323,27 @@ clean-support:
...
+
+define CLANG_ANALYZE
+clang-sa-$(1): $$(build_shlibdir)/libGCCheckerPlugin.$$(SHLIB_EXT)
+	@$$(call PRINT_ANALYZE, $$(build_depsbindir)/clang --analyze -Xanalyzer -analyzer-output=text -Xclang -load -Xclang $$(build_shlibdir)/libGCCheckerPlugin.$$(SHLIB_EXT) $$(CPPFLAGS) $$(CFLAGS) $$(DEBUGFLAGS) -Xclang -analyzer-checker=core$$(COMMA)julia.GCChecker --analyzer-no-default-checks -fcolor-diagnostics -Werror -x c $$(SRCDIR)/$(1).c)
+endef
```
Comment (vtjnash) [inline src/Makefile]:
```suggestion
.PHONY: clang-sa-$(1)
endef
```

File: src/Makefile [hunk 3]
```diff
@@ -323,13 +323,27 @@ clean-support:
...
 ifeq ($(USE_BINARYBUILDER_LLVM),0)
 ifneq ($(BUILD_LLVM_CLANG),1)
 	$(error Clang must be available to use the clang analyzer. Either build it (BUILD_LLVM_CLANG) or use BinaryBuilder)
 endif
 endif
-	$(MAKE) -C clangsa
-	$(build_depsbindir)/clang --analyze -Xanalyzer -analyzer-output=text -Xclang -load -Xclang $(build_shlibdir)/libGCCheckerPlugin.$(SHLIB_EXT)  $(CPPFLAGS) $(CFLAGS) $(DEBUGFLAGS) -Xclang -analyzer-checker=core,julia.GCChecker --analyzer-no-default-checks -fcolor-diagnostics -Werror -x c $(RUNTIME_C_SRCS:%=$(SRCDIR)/%.c)
+	$(MAKE) $(addprefix clang-sa-,$(RUNTIME_C_SRCS))
 
 .PHONY: default all debug release clean cleanall clean-* libccalltest libllvmcalltest julia_flisp.boot.inc.phony
```
Comment (vtjnash) [inline src/Makefile]:
```suggestion
.PHONY: default all debug release clean cleanall clean-* libccalltest libllvmcalltest julia_flisp.boot.inc.phony analyzegc
```

File: src/Makefile [hunk 4]
```diff
@@ -323,13 +323,27 @@ clean-support:
...
+#   make -c deps install-libuv install-utf8proc install-unwind
+analyzegc: $(BUILDDIR)/julia_version.h $(BUILDDIR)/julia_flisp.boot.inc $(build_shlibdir)/libGCCheckerPlugin.$(SHLIB_EXT)
 ifeq ($(USE_BINARYBUILDER_LLVM),0)
 ifneq ($(BUILD_LLVM_CLANG),1)
 	$(error Clang must be available to use the clang analyzer. Either build it (BUILD_LLVM_CLANG) or use BinaryBuilder)
 endif
 endif
-	$(MAKE) -C clangsa
-	$(build_depsbindir)/clang --analyze -Xanalyzer -analyzer-output=text -Xclang -load -Xclang $(build_shlibdir)/libGCCheckerPlugin.$(SHLIB_EXT)  $(CPPFLAGS) $(CFLAGS) $(DEBUGFLAGS) -Xclang -analyzer-checker=core,julia.GCChecker --analyzer-no-default-checks -fcolor-diagnostics -Werror -x c $(RUNTIME_C_SRCS:%=$(SRCDIR)/%.c)
+	$(MAKE) $(addprefix clang-sa-,$(RUNTIME_C_SRCS))
```
Comment (vtjnash) [inline src/Makefile]: recursive calls to the same Makefile tend to be a bad idea

File: src/flisp/Makefile [hunk 5]
```diff
@@ -59,9 +62,9 @@ $(BUILDDIR)/flisp.dbg.obj:  flisp.c cvalues.c types.c flisp.h print.c read.c equ
 $(BUILDDIR)/flmain.o:  flmain.c flisp.h
 $(BUILDDIR)/flmain.dbg.obj: flmain.c flisp.h
 
-$(LLT_release): $(LLTDIR)/*.h $(LLTDIR)/*.c
+$(LLT_release): $(LLTDIR)/*.h $(LLTDIR)/*.c flisp-deps
```
Comment (vtjnash) [inline src/flisp/Makefile]: It's invalid to depend on a `.PHONY` target

File: src/Makefile (hunk 4 above)
Comment (staticfloat) [reply src/Makefile]: Agreed; the need for this has decreased significantly now that I've eliminated the rest of the calls within this rule; I'll fix this to instead depend on a `clang-sa-deps-check` rule that conditionally runs the `$(error)`, then we can do everything within the same invocation.

File: src/flisp/Makefile (hunk 5 above)
Comment (staticfloat) [reply src/flisp/Makefile]: I'm not sure I would call it _invalid_, but it _does_ mean that `$(LLT_release)` will always be out of date, which isn't what I intended.

File: src/Makefile (hunk 1 above)
Comment (staticfloat) [reply src/Makefile]: @Keno can you confirm or deny?

Comment (Keno) [reply src/Makefile]: That'd probably work.

Comment (staticfloat) [reply src/Makefile]: I took at shot at this; let me know if I did something wrong; we're passing an awful lot more flags through now, but it seems to work from my perspective.

**Comment by staticfloat** [comment]
Triggered a new CI run; looks like it's working! :D

Will `clang` return nonzero if something fails an analysis sweep?  If not, it's always going to be green.

**Comment by Keno** [comment]
> Will `clang` return nonzero if something fails an analysis sweep? If not, it's always going to be green.

I think so, but you can try it out, but just deleting any `JL_GC_POP` line. The analyzer will catch that.

**Comment by staticfloat** [comment]
> I think so, but you can try it out, but just deleting any JL_GC_POP line. The analyzer will catch that.

[Looks like the answer is no](https://build.julialang.org/#/builders/146/builds/17/steps/2/logs/stdio), `clang` does not turn analyzer warnings into errors, even with `-Werror`. Here's [an SO answer](https://stackoverflow.com/questions/5033417/how-to-treat-warnings-from-clang-static-code-analysis-as-errors-in-xcode-3/5033483) about the same thing, with a very mac/iphone specific solution that I don't think will work for us.

We could check to make sure that `clang` does not output anything, but that seems a little fragile.

**Comment by StefanKarpinski** [comment]
Wut?! This is UNIX 101. Maybe submit a patch to `clang` to make it work the right way when running in static analysis mode?

**Comment by staticfloat** [comment]
@Keno is it possible for us to write a checker plugin that processes the results of all other checkers and turns any exception into something fatal?  The only other option that I can think of is to have it output something more structured, such as a `.plist`, consume that `.plist` and ensure that it is an empty array.

**Comment by Keno** [comment]
I looked into this for a bit, but didn't see anything obvious. Enabling the analyzer explicitly disables `-Werror` and the analyzer is hardcoded to emit its results as a warning. I do feel like we must be missing something stupid, so probably worth emailing the clang list to ask about.

**Comment by staticfloat** [comment]
I'm going to merge this for now, to get rid of the red X's, and because we will want these changes merged in any case; we'll fix the fault detection in a future PR.

**Comment by Keno** [comment]
Shall I email cfe-dev or do you want to?

**Comment by staticfloat** [comment]
Please go ahead; I couldn't even find the source for the analyzer parts of `clang`.

**Comment by Keno** [comment]
With https://reviews.llvm.org/D62885, adding `-Xanalyzer -analyzer-werror` will do what you want

**Pull request merged**